inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/hook/HookManager.java

```diff
@@ -0,0 +1,221 @@
+import rocks.inspectit.oce.bootstrap.instrumentation.noop.NoopHookManager;
+import rocks.inspectit.oce.bootstrap.instrumentation.noop.NoopMethodHook;
+import rocks.inspectit.oce.core.config.InspectitConfigChangedEvent;
+import rocks.inspectit.oce.core.config.InspectitEnvironment;
+import rocks.inspectit.oce.core.config.model.instrumentation.InternalSettings;
+import rocks.inspectit.oce.core.instrumentation.config.InstrumentationConfigurationResolver;
+import rocks.inspectit.oce.core.instrumentation.config.event.InstrumentationConfigurationChangedEvent;
+import rocks.inspectit.oce.core.instrumentation.config.model.MethodHookConfiguration;
+import rocks.inspectit.oce.core.instrumentation.event.IClassDiscoveryListener;
+import rocks.inspectit.oce.core.selfmonitoring.SelfMonitoringService;
+import rocks.inspectit.oce.core.service.BatchJobExecutorService;
+import rocks.inspectit.oce.core.utils.CommonUtils;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import java.lang.instrument.Instrumentation;
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Service
+public class HookManager implements IClassDiscoveryListener {
+
+    @Autowired
+    private BatchJobExecutorService executor;
+
+    @Autowired
+    private InspectitEnvironment env;
+
+    @Autowired
+    private Instrumentation instrumentation;
+
+    @Autowired
+    private InstrumentationConfigurationResolver configResolver;
+
+    @Autowired
+    private SelfMonitoringService selfMonitoring;
+
+    @Autowired
+    private MethodHookGenerator hookGenerator;
+
+    private LoadingCache<Class<?>, ConcurrentHashMap<String, MethodHook>> hooks = CacheBuilder.newBuilder().weakKeys().build(
+            new CacheLoader<Class<?>, ConcurrentHashMap<String, MethodHook>>() {
+                @Override
+                public ConcurrentHashMap<String, MethodHook> load(Class<?> key) throws Exception {
+                    return new ConcurrentHashMap<>();
+                }
+            });
+
+
+    /**
+     * The set of classes which might need hook updates.
+     * This service works through this set in batches.
+     */
+    @VisibleForTesting
+    Cache<Class<?>, Boolean> pendingClasses =
+            CacheBuilder.newBuilder().weakKeys().build();
+
+
+    private BatchJobExecutorService.BatchJob<Integer> classInstrumentationJob;
+
+    @PostConstruct
+    void init() {
+        Instances.hookManager = this::getHook;
+
+        InternalSettings conf = env.getCurrentConfig().getInstrumentation().getInternal();
+        Duration delay = conf.getInterBatchDelay();
+
+        classInstrumentationJob = executor.startJob(this::checkClassesForConfigurationUpdates, conf.getClassConfigurationCheckBatchSize(), delay, delay);
+    }
+
+    @PreDestroy
+    void destroy() {
```

Comment:
we are not clearing any caches on the destroy, I guess this is same as with the other similar classes, should we do that somehow?